dask/array/creation.py

```diff
@@ -143,3 +144,46 @@ def arange(*args, **kwargs):
         elem_count += bs
 
     return Array(dsk, name, chunks, dtype=dtype)
+
+
+def diag(v):
+    """Construct a diagonal array, with ``v`` on the diagonal.
+
+    Currently only implements diagonal array creation on the zeroth diagonal.
+    Support for the ``k``th diagonal or diagonal extraction, as per the numpy
+    interface, is not yet implemented.
+
+    Parameters
+    ----------
+    v : dask array
+
+    Returns
+    -------
+    out_array : dask array
+
+    Examples
+    --------
+
+    >>> diag(arange(3, chunks=3)).compute()
+    array([[0, 0, 0],
+           [0, 1, 0],
+           [0, 0, 2]])
+    """
+    if not isinstance(v, Array):
+        raise TypeError("v must be a dask array")
+    if v.ndim != 1:
+        raise NotImplementedError("Extracting diagonals with `diag` is not "
+                                  "implemented.")
+    chunks_1d = v.chunks[0]
+    name = next(diag_names)
+
+    blocks = v._keys()
+    dsk = v.dask.copy()
+    for i, m in enumerate(chunks_1d):
+        for j, n in enumerate(chunks_1d):
+            key = (name, i, j)
+            if i == j:
+                dsk[key] = (np.diag, blocks[i])
+            else:
+                dsk[key] = (np.zeros, (m, n))
+    return Array(dsk, name, (chunks_1d, chunks_1d), dtype=v.dtype)
```

Comment:
This should probably be `v._dtype` this is the dtype if it's known and None otherwise.  `v.dtype` is a property that is computed if `v._dtype is None`.